aten/src/ATen/native/sparse/SoftMax.cpp

```diff
@@ -0,0 +1,647 @@
+        data = A.to_dense().resize((nnz,))
+      then
+        data[offsets[n]] == A._values()[n]
+
+    `indices` must be a contiguous 2-d tensor with index_t entries.
+    `sizes` must be a vector with at least ndim entries.
+
+    `dim` is an integer. When >= 0 and < ndim, the indices of all
+    entries in the given dimension will be mapped to the index of the
+    first entry before computing the offset. Otherwise, the value is
+    ignored.
+
+    For example, consider a sparse tensor
+
+      11 ** ** 14 15
+      ** 22 ** 24 **
+
+    with
+
+      indices = [[0, 0, 0, 1, 1],
+                 [0, 3, 4, 1, 3]]
+
+    then
+
+      get_offsets(indices, (2, 5), -1) -> [0, 3, 4, 6, 8]
+      get_offsets(indices, (2, 5), 0) -> [0, 3, 4, 1, 3]
+      get_offsets(indices, (2, 5), 1) -> [0, 0, 0, 5, 5]
+
+  */
+  auto ndim = indices.size(0);
+  auto nnz = indices.size(1);
+  std::vector<index_t> offsets(nnz);
+  std::vector<index_t> strides(ndim, 1);
+  auto indices_accessor = indices.accessor<index_t, 2>();
+
+  if (ndim > 1) {
+    for (int64_t i=ndim - 2; i >= 0; i--) {
+      strides[i] = strides[i + 1] * sizes[i + 1];
+    }
+  }
+
+  for (int64_t i=0; i < nnz; i++) {
+    index_t acc = 0;
+    for (int64_t j=0; j < ndim; j++) {
+      auto indices_row = indices_accessor[j];
+      auto stride = strides[j];
+      if (j != dim) {
+        acc += stride * indices_row[i];
+      }
+    }
+    offsets[i] = acc;
+  }
+
+  return offsets;
+}
+
+template <typename index_t>
+std::vector<std::vector<index_t>> get_pools(const Tensor& indices, const std::vector<index_t>& sizes, const int64_t dim) {
+  /*
+    Return pools of indices that align with the given dimension.
+
+    Parameters:
+      `indices` - sparse tensor indices
+      `sizes`   - sparse tensor dimensions
+      `dim`     - given dimension
+
+    Returns:
+      `pools`   - vector of index vectors
+
+    A pool is defined as a list of indices (of sparse tensor values)
+    that participate in the same softmax computation:
+
+    - pools[i] intersection with pools[j] is empty iff i != j
+    - union of all pools is set(range(nnz))
+    - X.values[k], k in pools[i], does not affect the result of softmax(X)[n], n in pools[j], iff i != j
+
+  */
+
+  std::vector<std::vector<index_t>> pools;
+
+  auto ndim = indices.size(0);
+  auto nnz = indices.size(1);
+  std::vector<index_t> strides(ndim, 1);
+  auto indices_accessor = indices.accessor<index_t, 2>();
+
+  if (ndim > 1) {
+    for (int64_t i=ndim - 2; i >= 0; i--) {
+      strides[i] = strides[i + 1] * (i + 1 == dim? 1 : sizes[i + 1]);
+    }
+  }
+
+  for (int64_t i=0; i < nnz; i++) {
+    index_t pool_index = 0;
+    for (int64_t j=0; j < ndim; j++) {
+      if (j != dim) {
+        auto indices_row = indices_accessor[j];
+        auto stride = strides[j];
+        pool_index += stride * indices_row[i];
+      }
+    }
+    while (pool_index >= pools.size()) {
+      std::vector<index_t> indices;
+      pools.push_back(indices);
```

Comment:
I think `pools.push_back()` would be equivalent here.

Reply:
It is `pools.emplace_back()`. Fixed.